AGS 4: Separated Fonts and FontFiles, let freely change Font's source file and size

ivan-mogilko opened the pull request:
For #996

This is a preliminary draft; it is functional but there's still a doubt about choice of item arrangement in the project tree, as mentioned in the comments to #996, so that's potentially a subject to change.

Summary:
- Separate FontFile and Font items, stored and added each on their own.
- Font item's SourceFilename, which previously pointed to the original imported file, now works as a reference to a FontFile.
- Font's SourceFilename and Font Size may be freely modified in the editor. When modified, font is reloaded using specified settings. SourceFilename is a dropdown list, updated whenever a list of FontFiles changes, and Font Size is edited by pressing "..." button to the right, on which it displays a dialog, because there are 2 ways to set TTF size (this part remains from older versions).
- It is possible to have a Font without a source filename, in which case it won't draw anything, unless substituted by a plugin.
- FontFile properties are not viewable at this point, I'd have to make more changes for this, and also make a preview work for them.
- Standard font name "agsfntX.*" is deprecated, font files now may have any name. When compiling the game, Editor will package only files mentioned in the project tree.

Screenshots:
![ags4--newfonts1](https://github.com/user-attachments/assets/a16be39a-13dc-47c8-bfea-219fcef0fd95)
![ags4--newfonts2](https://github.com/user-attachments/assets/26ddba66-1d81-4e7b-91e7-5d96017ef915)
![ags4--newfonts3](https://github.com/user-attachments/assets/dd7bc40b-1916-4ea9-be4a-5f56e697de6f)

ivan-mogilko commented:
I suppose it will also help if there was some visual cue for each different kind of node... different icon, or text font or style used...

ivan-mogilko commented:
@AlanDrake could you please post the new icons that you made previously here, for easier reference?

AlanDrake commented:
![immagine](https://github.com/user-attachments/assets/69111937-17dd-4bd2-b17b-79197853ee00)

[newfonts.zip](https://github.com/user-attachments/files/17299027/newfonts.zip)

ivan-mogilko commented:
I think this is ready for review.

Other things that were mentioned in #996 were calling font items "Font Styles" and moving font files into "Fonts" subfolder on disk.

ericoporto commented on Editor/AGS.Editor/Components/FontsComponent.cs:
```diff
@@ -40,71 +53,63 @@ public override string ComponentID
+
+        private void ShowOrAddPane(FontFile chosenFontFile)
+        {
+
```
If this function is meant to be empty but needs to exist, I suggest adding a comment here, something like "panes only exist for Font for now, but we may call this method for a node that may be a FontFile, just do nothing".

ericoporto commented:
I am not a super fan of not renaming the font files, I see a few `EndsWith(".ttf")` and similar, these all need to be case insensitive if we are keeping the original font name.

ivan-mogilko commented:
> I am not a super fan of not renaming the font files, I see a few `EndsWith(".ttf")` and similar, these all need to be case insensitive if we are keeping the original font name. If we renamed on import then these would be guaranteed to work in case sensitive.

Please elaborate, are there other arguments against not keeping original name, or just that I used incorrect comparison function in code? When you think that they should be renamed, do you mean only the letter case, or you think that font files should keep old names "agsfntN.ext"?

Personally, I try to assume that the game may be constructed by gathering the files by hand, without use of editor.

ivan-mogilko commented:
> I see a few `EndsWith(".ttf")` and similar, these all need to be case insensitive if we are keeping the original font name. If we renamed on import then these would be guaranteed to work in case sensitive.

BTW, I searched for `EndsWith(".` and there are numerous cases like this throughout the Editor code.
I suppose that we need to replace these with a helper function, like TestExtension, that compares case insensitive.

ivan-mogilko commented:
Per @AlanDrake 's suggestion (posted elsewhere, maybe on Discord?), changed font file location to a "Fonts" subdirectory.

Fonts are copied over on project upgrade, and Editor even makes a folder backup if it existed previously, complying to how RoomComponent does this when upgrading rooms.

I had to modify "--runfromide" engine's command, because fonts are not packed when running a test game, editor must tell an extra directory to the engine. Existing arg list for "runfromide" is rather inconvenient, with hardcoded meaning for each entry. So I modified it to be a custom list of dirs paired with "asset filter" list (asset filters are used to narrow search of certain asset types).
